### PR TITLE
feat: autodiscover repos by topic

### PR DESCRIPTION
**Pull request by dhduvall**
## Changes

This creates a new global configuration option, `autodiscoverTopics`, which allows the renovate admin to autodiscover repos by looking at their topics. This PR only implements support for GitLab.

## Context

This was requested in discussion https://github.com/renovatebot/renovate/issues/10011.

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/platform/types.ts [hunk 1]
```diff
@@ -163,6 +163,10 @@ export type EnsureCommentRemovalConfig =
 
 export type EnsureIssueResult = 'updated' | 'created';
 
+export interface AutodiscoverParams {
```
Comment (viceice) [inline lib/modules/platform/types.ts:166]:
```suggestion
export interface AutodiscoverConfig {
```
we mostly use config as name

File: lib/modules/platform/gitlab/index.ts [hunk 2]
```diff
@@ -142,10 +143,23 @@ export async function initPlatform({
 }
 
 // Get all repositories that the user has access to
-export async function getRepos(): Promise<string[]> {
+export async function getRepos(params?: AutodiscoverParams): Promise<string[]> {
```
Comment (viceice) [inline lib/modules/platform/gitlab/index.ts:146]:
```suggestion
export async function getRepos(confug?: AutodiscoverConfig): Promise<string[]> {
```

File: lib/modules/platform/gitlab/index.ts [hunk 3]
```diff
@@ -142,10 +143,23 @@ export async function initPlatform({
...
   logger.debug('Autodiscovering GitLab repositories');
+
+  let topicString = '';
+  if (params?.topics?.length) {
```
Comment (viceice) [inline lib/modules/platform/gitlab/index.ts:150]:
```suggestion
  if (config?.topics?.length) {
```

File: lib/modules/platform/types.ts [hunk 4]
```diff
@@ -190,7 +194,7 @@ export interface Platform {
   addReviewers(number: number, reviewers: string[]): Promise<void>;
   addAssignees(number: number, assignees: string[]): Promise<void>;
   createPr(prConfig: CreatePRConfig): Promise<Pr | null>;
-  getRepos(): Promise<string[]>;
+  getRepos(param?: AutodiscoverParams): Promise<string[]>;
```
Comment (viceice) [inline lib/modules/platform/types.ts:197]:
```suggestion
  getRepos(config?: AutodiscoverConfig): Promise<string[]>;
```

File: lib/modules/platform/gitlab/index.ts [hunk 5]
```diff
@@ -31,6 +31,7 @@ import {
 } from '../../../util/url';
 import { getPrBodyStruct } from '../pr-body';
 import type {
+  AutodiscoverParams,
```
Comment (viceice) [inline lib/modules/platform/gitlab/index.ts:34]:
```suggestion
  AutodiscoverConfig,
```

File: lib/modules/platform/gitlab/index.ts [hunk 6]
```diff
@@ -142,10 +143,23 @@ export async function initPlatform({
 }
 
 // Get all repositories that the user has access to
-export async function getRepos(): Promise<string[]> {
+export async function getRepos(params?: AutodiscoverParams): Promise<string[]> {
   logger.debug('Autodiscovering GitLab repositories');
+
+  let topicString = '';
+  if (params?.topics?.length) {
+    topicString = `&topic=${encodeURIComponent(params.topics.join(','))}`;
```
Comment (viceice) [inline lib/modules/platform/gitlab/index.ts:151]:
```suggestion
    topicString = `&topic=${encodeURIComponent(config.topics.join(','))}`;
```

File: lib/modules/platform/gitlab/index.ts [hunk 7]
```diff
@@ -142,10 +143,23 @@ export async function initPlatform({
...
+  let topicString = '';
+  if (config?.topics?.length) {
+    topicString = `&topic=${encodeURIComponent(config.topics.join(','))}`;
+  }
+
   try {
-    const url = `projects?membership=true&per_page=100&with_merge_requests_enabled=true&min_access_level=30&archived=false`;
+    const url =
+      'projects' +
+      '?membership=true' +
```
Comment (viceice) [inline lib/modules/platform/gitlab/index.ts:157]: we should probably use the urlparams class here instead. it makes param handling simpler

File: lib/modules/platform/types.ts [hunk 8]
```diff
@@ -190,7 +194,7 @@ export interface Platform {
   addReviewers(number: number, reviewers: string[]): Promise<void>;
   addAssignees(number: number, assignees: string[]): Promise<void>;
   createPr(prConfig: CreatePRConfig): Promise<Pr | null>;
-  getRepos(): Promise<string[]>;
+  getRepos(param?: AutodiscoverConfig): Promise<string[]>;
```
Comment (viceice) [inline lib/modules/platform/types.ts:197]:
```suggestion
  getRepos(config?: AutodiscoverConfig): Promise<string[]>;
```
whoops, wrong suggestion

File: lib/modules/platform/gitlab/index.ts (hunk 7 above)
Comment (dhduvall) [reply lib/modules/platform/gitlab/index.ts:157]: Oh, much better; thanks.

File: lib/modules/platform/types.ts (hunk 8 above)
Comment (dhduvall) [reply lib/modules/platform/types.ts:197]: That was just me missing that instance of "param".

**Review by viceice — APPROVED** [review]

**Pull request merged**